File: jmolecules-axon/02-onion-cqrs-es-bank/src/main/java/org/jmolecules/example/axonframework/domain/api/exception/InsufficientBalanceException.java
package org.jmolecules.example.axonframework.domain.api.exception;


public class InsufficientBalanceException extends RuntimeException{

  public InsufficientBalanceException(String accountId, int currentBalance, int withdrawAmount, int minAmount) {
    super(String.format("BankAccount[id=%s, currentBalance=%d]: Withdrawal of amount=%d not allowed, would subceed min. balance of %d",
                        accountId,
                        currentBalance,
                        withdrawAmount,
                        minAmount
    ));
  }
}

File: jmolecules-axon/03-onion-axon-cqrs-es-bank/src/main/kotlin/org/jmolecules/example/axonframework/domain/api/exception/InsufficientBalanceException.kt
package org.jmolecules.example.axonframework.domain.api.exception

import org.jmolecules.example.axonframework.domain.api.type.AccountId
import org.jmolecules.example.axonframework.domain.api.type.Amount
import org.jmolecules.example.axonframework.domain.api.type.Balance

class InsufficientBalanceException(
    accountId: AccountId,
    currentBalance: Balance,
    withdrawAmount: Amount?,
    minimumBalance: Balance
) : RuntimeException(
    if (withdrawAmount == null) {
        "BankAccount[id=$accountId] can't be created, its balance $currentBalance would subceed min. balance of $minimumBalance"
    } else {
        "BankAccount[id=$accountId, currentBalance=$currentBalance]: Withdrawal of amount=$withdrawAmount not allowed, would subceed min. balance of $minimumBalance"
    }
)
